### PR TITLE
Use cmake

### PR DESCRIPTION
**Pull request by BusyJay**
This pr changes the way compiling and linking to rocksdb.

In the past, rocksdb is linked dynamically by default. When link to rocksdb statically, shell script is used to download and build all the native libraries. This method is straightforward and worked quite well.

However, things change now. There are several patches and bug fixes that are not merged into upstream yet (and some are probably not merged at all), if we are stick to the old way, there will be conflict between our fork and official release.

Besides, dynamic link is hard to use and maintain. Many issues in TiKV are related to linking to rocksdb dynamically. And shell script is not platform-independent.

So this pr makes rocksdb as a subtree of the project, so that it doesn't have to be downloaded again and again very time clean is run. And then use cmake to skip the details of compilation. Dynamic link is disabled now, only the subtree version of rocksdb is supported.

### CODE REVIEW

**Comment by siddontang** [comment]
do we need to use submodule?

**Comment by BusyJay** [comment]
No, use subtree is OK. When using submodule, more than 50 Mib data needs to be downloaded. However, when using subtree, less than 6 Mib data is required.

**Comment by siddontang** [comment]
@BusyJay 

Do we have a better way to update RocksDB from our forked repository later?

**Comment by BusyJay** [comment]
Ping, do you have any concern about it?

**Comment by zhangjinpeng1987** [comment]
@BusyJay  Can you give an explanation how to use it?

**Comment by siddontang** [comment]
can we prune the RocksDB source, E.g, only keep the source file and build scripts.

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM

File: librocksdb_sys/Cargo.toml [hunk 1]
```diff
@@ -3,18 +3,37 @@ name = "librocksdb_sys"
...
+[dependencies.libz-sys]
+git = "https://github.com/busyjay/libz-sys.git"
+branch = "static-link"
+
+[dependencies.bzip2-sys]
+git = "https://github.com/alexcrichton/bzip2-rs.git"
+
+[dependencies.lz4-sys]
+git = "https://github.com/busyjay/lz4-rs.git"
+branch = "adjust-build"
```
Comment (zhangjinpeng1987) [inline librocksdb_sys/Cargo.toml]: What does this branch mean?

Comment (BusyJay) [reply librocksdb_sys/Cargo.toml]: Enable static link support. /cc bozaro/lz4-rs#23

File: librocksdb_sys/src/lib.rs [hunk 2]
```diff
@@ -16,6 +16,7 @@
 extern crate libc;
 #[cfg(test)]
 extern crate tempdir;
+extern crate bzip2_sys;
```
Comment (huachaohuang) [inline librocksdb_sys/src/lib.rs]: What's this for?

File: tests/test_iterator.rs [hunk 3]
```diff
@@ -170,8 +170,8 @@ pub fn test_iterator() {
     iter.next();
     assert!(!iter.valid());
     // Once iterator is invalid, it can't be reverted.
-    iter.prev();
-    assert!(!iter.valid());
+    //iter.prev();
```
Comment (huachaohuang) [inline tests/test_iterator.rs]: Why comment out this two lines?

File: librocksdb_sys/src/lib.rs (hunk 2 above)
Comment (BusyJay) [reply librocksdb_sys/src/lib.rs]: bzip2 is compiled with stdio disabled, so there has to be an external hook to handle error, which is defined in bzip2_sys already.

File: tests/test_iterator.rs (hunk 3 above)
Comment (BusyJay) [reply tests/test_iterator.rs]: Because when iterator is not valid anymore, `prev` should not be called, otherwise assertion in prev will fail.

**Comment by BusyJay** [comment]
Windows support is not complete yet, there are still some issues on windows platform. But I think it's fine to merge it first, and deal with the issues later.

**Pull request merged**